bundles/admin/admin-layereditor/instance.js

```diff
@@ -168,5 +237,86 @@ Oskari.clazz.defineES('Oskari.admin.admin-layereditor.instance',
             }
             return this.flyout;
         }
+
+        /**
+         * @method _getAddThemeFlyout
+         * Ensures theme flyout exists and returns it.
+         * @return {LocalizingFlyout}
+         */
+        _getAddThemeFlyout () {
+            if (this.themeFlyout) {
+                return this.themeFlyout;
+            }
+            this.themeFlyout = new LocalizingFlyout(this, this.loc('addTheme'), {
+                showHeading: false,
+                headerMessageKey: 'themeName'
+            });
+            this.themeFlyout.makeDraggable({
+                handle: '.oskari-flyouttoolbar',
+                scroll: false
+            });
+            this.themeFlyout.setAction(value => {
+                // TODO add discreet notifications
+                jQuery.ajax({
+                    type: 'PUT',
+                    dataType: 'json',
+                    contentType: 'application/json',
+                    url: Oskari.urls.getRoute('MapLayerGroups'),
+                    data: JSON.stringify({ locales: value }),
+                    success: response => {
+                        this.themeFlyout.hide();
+                        const group = Oskari.clazz.create('Oskari.mapframework.domain.MaplayerGroup', response);
+                        this._getLayerService().addLayerGroup(group);
+                    },
+                    error: () => {
+                        this.themeFlyout.setLoading(false);
+                    }
+                });
+            });
+            return this.themeFlyout;
+        }
+
+        /**
+         * @method _getAddDataProviderFlyout
+         * Ensures theme flyout exists and returns it.
+         * @return {LocalizingFlyout}
+         */
+        _getAddDataProviderFlyout () {
+            if (this.dataProviderFlyout) {
+                return this.dataProviderFlyout;
+            }
+            this.dataProviderFlyout = new LocalizingFlyout(this, this.loc('addDataProvider'), {
+                showHeading: false,
+                headerMessageKey: 'dataProviderName'
+            });
+            this.dataProviderFlyout.makeDraggable({
+                handle: '.oskari-flyouttoolbar',
+                scroll: false
+            });
+            this.dataProviderFlyout.setAction(value => {
+                // TODO add discreet notifications
+                jQuery.ajax({
+                    type: 'PUT',
```

Comment:
Currently "action" is only for adding? I think we need some modifications for making editing work. Also we might want to add delete to the popup so this could be setSaveAction() or something like that?

Reply:
Indeed, the flyout can perform only single action at this point. Refactoring is required once we implement editing and deleting data.

Reply:
Renamed `setAction` to `setSaveAction`.